File: pos_credit_note_gift_card/models/customer_account_return_guard.py
# -*- coding: utf-8 -*-
"""
customer_account_return_guard.py
================================
Server-side guard for credit-note returns of Customer Account sales.

Business rule
-------------
If the original POS order was paid with Customer Account / credit
(``pos.payment.method.type == 'pay_later'`` or the optional
``pcl_is_credit_method`` flag), a cashier must not create a gift-card
credit note for the return. The return should be completed through the
Customer Account payment method so the customer's receivable balance is
reduced.
"""

import logging

from odoo import _, api, models

_logger = logging.getLogger("pos_credit_note_gift_card.customer_account_return_guard")


class PosOrderCustomerAccountReturnGuard(models.Model):
    _inherit = "pos.order"

    @api.model
    def check_customer_account_return_block(self, original_order_ids):
        """Return a structured block result for original sale order ids."""
        clean_ids = []
        for order_id in original_order_ids or []:
            try:
                order_id = int(order_id)
            except (TypeError, ValueError):
                _logger.warning(
                    "[CustomerAccountReturnGuard] Ignoring invalid original_order_id=%r",
                    order_id,
                )
                continue
            if order_id:
                clean_ids.append(order_id)

        orders = self.browse(clean_ids).exists()
        blocked_orders = []

        _logger.info(
            "[CustomerAccountReturnGuard] Checking original_order_ids=%s resolved=%s",
            clean_ids,
            orders.ids,
        )

        for order in orders:
            credit_payments = order.payment_ids.filtered(
                lambda payment: self._cn_is_customer_account_payment_method(
                    payment.payment_method_id
                )
            )
            payment_debug = [
                {
                    "payment_id": payment.id,
                    "method_id": payment.payment_method_id.id,
                    "method_name": payment.payment_method_id.name,
                    "method_type": payment.payment_method_id.type,
                    "pcl_is_credit_method": bool(
                        getattr(payment.payment_method_id, "pcl_is_credit_method", False)
                    ),
                    "amount": payment.amount,
                }
                for payment in order.payment_ids
            ]
            _logger.debug(
                "[CustomerAccountReturnGuard] order=%s id=%s payment_debug=%s",
                order.name,
                order.id,
                payment_debug,
            )

            if credit_payments:
                method_names = ", ".join(credit_payments.mapped("payment_method_id.name"))
                blocked_orders.append(
                    {
                        "id": order.id,
                        "name": order.name,
                        "payment_methods": method_names,
                    }
                )
                _logger.warning(
                    "[CustomerAccountReturnGuard] BLOCK original order '%s' id=%s "
                    "paid by Customer Account methods=%s",
                    order.name,
                    order.id,
                    method_names,
                )

        if not blocked_orders:
            return {
                "blocked": False,
                "message": "",
                "orders": [],
            }

        order_names = ", ".join(order["name"] for order in blocked_orders)
        message = _(
            "Credit note generation is blocked because the original POS order "
            "%(orders)s was paid by Customer Account. Complete this return using "
            "the Customer Account payment method so the customer's account balance "
            "is corrected.",
            orders=order_names,
        )
        return {
            "blocked": True,
            "message": message,
            "orders": blocked_orders,
        }

    @api.model
    def _cn_is_customer_account_payment_method(self, payment_method):
        """True for Odoo Customer Account / optional pos_credit_limit marker."""
        if not payment_method:
            return False
        return bool(
            payment_method.type == "pay_later"
            or getattr(payment_method, "pcl_is_credit_method", False)
        )

    def validate_return_for_credit_note(self, config_id):
        """Append the Customer Account block to the existing validation result."""
        result = super().validate_return_for_credit_note(config_id)
        for order in self:
            original_order_ids = order.lines.mapped("refunded_orderline_id.order_id").ids
            guard = self.check_customer_account_return_block(original_order_ids)
            if guard.get("blocked"):
                result.setdefault("errors", []).append(guard["message"])
                result["ok"] = False
                _logger.warning(
                    "[CustomerAccountReturnGuard] validate_return_for_credit_note blocked "
                    "return_order=%s original_order_ids=%s",
                    order.name,
                    original_order_ids,
                )
        return result

File: pos_credit_note_gift_card/static/src/js/customer_account_return_guard.js
/** @odoo-module **/
/**
 * Blocks credit-note generation for returns whose original POS order was paid
 * using Customer Account / credit. Logs are always visible in Chrome DevTools
 * with prefix [CNCustomerAccountReturnGuard].
 */

import { patch } from "@web/core/utils/patch";
import { TicketScreen } from "@point_of_sale/app/screens/ticket_screen/ticket_screen";
import { PaymentScreen } from "@point_of_sale/app/screens/payment_screen/payment_screen";
import { useService } from "@web/core/utils/hooks";
import { _t } from "@web/core/l10n/translation";
import { AlertDialog } from "@web/core/confirmation_dialog/confirmation_dialog";
import { rpc } from "@web/core/network/rpc";

const LOG_PREFIX = "[CNCustomerAccountReturnGuard]";

function log(...args) {
    console.warn(LOG_PREFIX, ...args);
}

function debug(...args) {
    if (window.CN_DEBUG) {
        console.debug(LOG_PREFIX, ...args);
    }
}

function paymentMethodIsCustomerAccount(paymentMethod) {
    return Boolean(
        paymentMethod &&
        (paymentMethod.type === "pay_later" || paymentMethod.pcl_is_credit_method)
    );
}

function orderPaidWithCustomerAccount(order) {
    const payments = order?.payment_ids || [];
    const creditPayments = payments.filter((payment) =>
        paymentMethodIsCustomerAccount(payment.payment_method_id)
    );
    debug("orderPaidWithCustomerAccount", {
        order: order?.name,
        orderId: order?.id,
        serverId: order?.server_id,
        payments: payments.map((payment) => ({
            amount: payment.amount,
            methodId: payment.payment_method_id?.id,
            methodName: payment.payment_method_id?.name,
            methodType: payment.payment_method_id?.type,
            pclIsCredit: payment.payment_method_id?.pcl_is_credit_method,
        })),
        creditCount: creditPayments.length,
    });
    return creditPayments.length > 0;
}

function getOriginalOrdersFromRefundOrder(refundOrder) {
    const seen = new Set();
    const orders = [];
    for (const line of refundOrder?.lines || []) {
        const refLine = line.refunded_orderline_id;
        const originalOrder = refLine && typeof refLine === "object"
            ? refLine.order_id
            : null;
        const originalId = originalOrder && typeof originalOrder === "object"
            ? (originalOrder.id ?? originalOrder.server_id)
            : originalOrder;
        if (originalOrder && typeof originalOrder === "object" && !seen.has(originalId)) {
            seen.add(originalId);
            orders.push(originalOrder);
        }
    }
    return orders;
}

function getOriginalOrderIdsFromRefundOrder(refundOrder) {
    const seen = new Set();
    const ids = [];

    const returnRef = refundOrder?.return_pos_order_id;
    const returnRefId = returnRef && typeof returnRef === "object"
        ? (returnRef.server_id ?? returnRef.id)
        : returnRef;
    if (typeof returnRefId === "number") {
        seen.add(returnRefId);
        ids.push(returnRefId);
    }

    for (const line of refundOrder?.lines || []) {
        const refLine = line.refunded_orderline_id;
        const originalOrder = refLine && typeof refLine === "object"
            ? refLine.order_id
            : null;
        const rawId = originalOrder && typeof originalOrder === "object"
            ? (originalOrder.server_id ?? originalOrder.id)
            : originalOrder;
        if (typeof rawId === "number" && !seen.has(rawId)) {
            seen.add(rawId);
            ids.push(rawId);
        }
    }
    return ids;
}

function getSelectedTicketOriginalOrder(screen) {
    const order = screen.getSelectedOrder?.();
    const selectedId = screen.getSelectedOrderlineId?.();
    const selectedLine = selectedId
        ? order?.lines?.find((line) => line.id == selectedId)
        : null;
    const lineOrder = selectedLine?.order_id;
    if (lineOrder && typeof lineOrder === "object") {
        return lineOrder;
    }
    return order || null;
}

function blockMessage(orderName = "") {
    return _t(
        "Credit note generation is blocked for this return because the original POS order %(order)s was paid by Customer Account.\n\nUse Customer Account to complete the return.",
        { order: orderName || "" }
    );
}

async function callServerGuard(originalOrderIds) {
    if (!originalOrderIds.length) {
        log("Server guard skipped: no original order ids found.");
        return { blocked: false, message: "", orders: [] };
    }
    log("Calling server guard for originalOrderIds=", originalOrderIds);
    const response = await rpc("/pos/credit_note/customer_account_return_guard", {
        original_order_ids: originalOrderIds,
    });
    if (!response?.ok) {
        throw new Error(response?.error || "Customer Account return guard failed.");
    }
    log("Server guard response=", response.payload);
    return response.payload || { blocked: false, message: "", orders: [] };
}

async function getCustomerAccountBlock(refundOrder) {
    const localOriginalOrders = getOriginalOrdersFromRefundOrder(refundOrder);
    const localBlocked = localOriginalOrders.filter(orderPaidWithCustomerAccount);
    if (localBlocked.length) {
        const names = localBlocked.map((order) => order.name || order.pos_reference || order.id);
        log("LOCAL BLOCK: original order paid with Customer Account:", names);
        return {
            blocked: true,
            message: blockMessage(names.join(", ")),
            orders: localBlocked,
            source: "local",
        };
    }

    const originalOrderIds = getOriginalOrderIdsFromRefundOrder(refundOrder);
    try {
        const serverGuard = await callServerGuard(originalOrderIds);
        if (serverGuard.blocked) {
            return { ...serverGuard, source: "server" };
        }
    } catch (err) {
        console.error(LOG_PREFIX, "Server guard error:", err);
        throw err;
    }

    log("ALLOW: no Customer Account payment found on original order.", {
        refundOrder: refundOrder?.name,
        originalOrderIds,
    });
    return { blocked: false, message: "", orders: [], source: "none" };
}

patch(TicketScreen.prototype, {
    setup() {
        super.setup(...arguments);
        if (!this.dialog) {
            this.dialog = useService("dialog");
        }
        log("TicketScreen guard active.");
    },

    async _onDoRefund() {
        const originalOrder = getSelectedTicketOriginalOrder(this);
        if (originalOrder && orderPaidWithCustomerAccount(originalOrder)) {
            log("BLOCK at TicketScreen _onDoRefund", {
                originalOrder: originalOrder.name,
                originalOrderId: originalOrder.id,
            });
            this.dialog.add(AlertDialog, {
                title: _t("Use Customer Account for Return"),
                body: blockMessage(originalOrder.name || originalOrder.id),
            });
            return;
        }
        return super._onDoRefund?.(...arguments);
    },
});

patch(PaymentScreen.prototype, {
    setup() {
        super.setup(...arguments);
        log("PaymentScreen guard active.");
    },

    async _runCreditNoteCore(order, orderId, configId) {
        const guard = await getCustomerAccountBlock(order);
        if (guard.blocked) {
            log("BLOCK before credit-note core", {
                source: guard.source,
                orderId,
                configId,
                guard,
            });
            this.dialog.add(AlertDialog, {
                title: _t("Use Customer Account for Return"),
                body: guard.message || blockMessage(),
            });
            return null;
        }
        return super._runCreditNoteCore(...arguments);
    },
});
